Native client library audio mixer APIs.

namark opened the pull request:
aims to resolve https://github.com/vircadia/vircadia/issues/1643

As with avatar APIs the refactoring is partial, there is a new library consisting of parts of `audio-client` that were not interface specific, but it's not yet used in `audio-client` library itself, leaving that for later again in favor of quicker Unity SDK features.

There are conflicts between this and https://github.com/vircadia/vircadia/pull/1582, since I have separated codecs from plugins as part of refactoring.

namark committed:
Eliminate Transfrom dependency in AbstractAudioInterface,

no member functions are used, and the two calls construct a transform
object in place.

namark committed:
Eliminate AABox dependency in PositionalAudioStream,

replaced by a simpler structure and a function.

namark committed:
Eliminate dependency on ReginsterMetaTypes in AudioInjectorOptions,

by moving the script value conversion functions into
AudioScriptingInterface.

namark committed:
Eliminate CodecPlugin dependency in InboundAudioStream,

by separating the codec interface from the plugin interface.

namark committed:
Audio library no longer depends on GUI and plugins.

namark committed:
Moved opus and pcm codec implementations to libraries,

for future use in client library. The respective plugins are now just
    wrappers for these.

namark committed:
New library for audio client packet handling and audio format conversion,

based on audio-client library with interface specific parts removed.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-_76e62d8a.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-.tar_76e62d8a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-_76e62d8a.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-.tar_76e62d8a.Z

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-_76e62d8a.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-.tar_76e62d8a.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-.tar_76e62d8a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-80bf036-_76e62d8a.sh

namark committed:
Fixed some silly compiler errors.

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-.tar_c5922574.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-.tar_c5922574.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-_c5922574.sh

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-_c5922574.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-.tar_c5922574.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-.tar_c5922574.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-49590ca-_c5922574.sh

namark committed:
Microphone input sample type conversion for client library.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-Interface-PR1675-c320e04-_91aa0c5d.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-.tar_91aa0c5d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-.tar_91aa0c5d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-_91aa0c5d.sh

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-_91aa0c5d.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-.tar_91aa0c5d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-_91aa0c5d.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-c320e04-.tar_91aa0c5d.Z

namark committed:
Another silly compilation error.

namark committed:
AudioPacketHandler codec list interface.

namark committed:
Handling codec negotiation on mixer activation in AudioPacketHandler.

vircadia-build-notifier commented:
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-Interface-PR1675-e75d5a3-_ef8b5f1a.dmg

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-_ef8b5f1a.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-_ef8b5f1a.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-.tar_ef8b5f1a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-.tar_ef8b5f1a.Z

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-_ef8b5f1a.exe

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-.tar_ef8b5f1a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-_ef8b5f1a.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1675/Vircadia-PR1675-e75d5a3-.tar_ef8b5f1a.Z

namark committed:
Removed unused forward declaration that caused ambiguity.

namark committed:
Initial implementation of client library audio mixer API.

namark committed:
Replaced QAudioFormat with simple struct in audio-client-core library,

as it's the only dependency from QtMultimedia.

namark committed:
Fixed AudioFormat class qDebug output formatting.

namark committed:
Fixed unintended implicit sample type conversion in AudioPacketHandler.

namark committed:
Fixed audio output initialization check in AudioPacketHandler.

namark committed:
Moved AudioPacketHandler signal connections and handler registration into the audio thread.

namark committed:
Fixed native client library audio format validation logic.

namark committed:
Attempt to fix client library NodeList not being destroyed properly,

reference held somewhere.

namark committed:
Moved native client library AudioClient update timer code into audio thread.

namark committed:
Fixed client library audio dummy input being continuously set,

until a valid input was set by the user.

namark committed:
Fixed a forward declaration of AudioFormat to match the definition.

namark committed:
Removed an unnecessary include.

namark committed:
Improved AudioFormat interface to be less confusing.

namark committed:
Fixed audio output requested sample count calculation in AudioPacketHandler.

namark committed:
Client library vircadia_get_audio_output_data function now returns bytes written.

namark committed:
Minor typo fix.

namark committed:
Client library unit test for sending and receiving audio.

namark committed:
Validating audio input and output data buffers in client library.

namark committed:
Client library audio API documentation.

namark committed:
Fixed native client library audio unit test failing without a running server.

namark committed:
Fixed audio packet handler include guard to encompass includes.

namark committed:
CMake option for WebRTC audio processing.

namark committed:
Workaround Apple Clang bug with lambdas in templates,

it gives both default getter lambdas the same mangled name,
likely https://bugs.llvm.org/show_bug.cgi?id=33924,
XCode 11.2

namark committed:
Fixed an include guard.
